### PR TITLE
Use async functions instead of constructing Futures manually

### PR DESCRIPTION
**Pull request by greglittlefield-wf**
## Motivation
There were some `Future.error` getting constructed without a stacktrace, which resulted in no stacktrace being output in logs.

## Changes
Use async functions and throw errors instead of constructing Futures manually, which results in errors having stacktraces, and also cleans things up a bit.

#### Release Notes
- Fix missing stacktraces in most `StateError`s thrown by `LifecycleModule`

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

FYI @laurameckley-wk 

### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Anything falling under manual testing criteria [outlined in CONTRIBUTING.md][contributing-manual-testing]

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Client Platform member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/w_module/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/w_module/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

**Comment by aviary2-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://dev.workiva.net/docs/teams/information-security/aviary#configuring-aviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: lib/src/lifecycle_module.dart [hunk 1]
```diff
@@ -487,83 +488,79 @@ abstract class LifecycleModule extends SimpleModule with Disposable {
...
     }
 
     if (isUnloaded || isUnloading) {
       var stateLabel = isUnloaded ? 'unloaded' : 'unloading';
-      return Future.error(
-          StateError('Cannot load child module when module is $stateLabel'));
+      throw StateError('Cannot load child module when module is $stateLabel');
     }
 
-    final completer = Completer<Null>();
```
Comment (greglittlefield-wf) [inline lib/src/lifecycle_module.dart:505]: Best viewed with whitespace-agnostic diff: https://github.com/Workiva/w_module/pull/165/files?w=1#diff-c586827f6717f60db651d466b5ab2b5fL505

File: lib/src/lifecycle_module.dart [hunk 2]
```diff
@@ -487,83 +488,79 @@ abstract class LifecycleModule extends SimpleModule with Disposable {
...
-        completer.completeError(error, stackTrace);
-      } finally {
-        childModule._parentContext = null;
-      }
-    }).catchError((Object error, StackTrace stackTrace) {
+    try {
+      await onWillLoadChildModule(childModule);
+    } catch (error, stackTrace) {
       _logger.severe(
         'Exception in onWillLoadChildModule ($name)',
```
Comment (greglittlefield-wf) [inline lib/src/lifecycle_module.dart:509]: Wow having this error handling right next to the actual code that triggered the error is so much easier to follow

**Comment by greglittlefield-wf** [comment]
Soo it looks like this throws off the timing of how events are emitted, which I do NOT want to mess with. Closing in favor of a less-invasive fix

**Pull request closed without merging**